TeuchosKokkosCompat: Test fails to build

@trilinos/teuchos @trilinos/kokkos @trilinos/tpetra 

A test in TeuchosKokkosCompat fails to build.  I get a few errors like the following:
```
.../Trilinos/packages/teuchos/kokkoscompat/test/linkTest.cpp: In member function ‘virtual void LinkTeuchosAndKokkos_NoInteraction_UnitTest::runUnitTestImpl(Teuchos::FancyOStream&, bool&) const’:
.../Trilinos/packages/teuchos/kokkoscompat/test/linkTest.cpp:109:3: error: ‘initialize’ is not a member of ‘TestDevice {aka Kokkos::OpenMP}’
   TestDevice::initialize();
   ^
.../Trilinos/packages/teuchos/kokkoscompat/test/linkTest.cpp:126:3: error: ‘finalize’ is not a member of ‘TestDevice {aka Kokkos::OpenMP}’
   TestDevice::finalize();
   ^
```

## Related Issues
* Blocks #4249, #4234

--- a/packages/teuchos/core/src/Teuchos_ArrayView.hpp
+++ b/packages/teuchos/core/src/Teuchos_ArrayView.hpp
@@ -72,10 +72,20 @@ ArrayView<const T>::ArrayView( ENull )
 
 template<class T> inline
 ArrayView<T>::ArrayView( T* p, size_type size_in, const ERCPNodeLookup rcpNodeLookup )
-  :ptr_(p), size_(size_in)
+  :ptr_(size_in == 0 ? nullptr : p), size_(size_in)
 {
 #ifdef HAVE_TEUCHOS_ARRAY_BOUNDSCHECK
-  TEUCHOS_TEST_FOR_EXCEPT( p != 0 && size_in <= 0 );
+  // We comment out the one check below, as part of the fix for #4234:
+  //
+  // https://github.com/trilinos/Trilinos/issues/4234
+  //
+  // This permits conversion from std::vector or Kokkos::View, using
+  // ArrayView(x.data(), x.size(), RCP_DISABLE_NODE_LOOKUP).  The
+  // other part of the fix is that we make sure ptr_ is null if
+  // size_in is zero.
+  //
+  //TEUCHOS_TEST_FOR_EXCEPT( p != 0 && size_in <= 0 );
+
   TEUCHOS_TEST_FOR_EXCEPT( p == 0 && size_in != 0 );
   // This only does something if HAVE_TEUCHOS_ARRAY_BOUNDSCHECK is defined.
   setUpIterators(rcpNodeLookup);
@@ -86,10 +96,20 @@ ArrayView<T>::ArrayView( T* p, size_type size_in, const ERCPNodeLookup rcpNodeLo
 
 template<class T> inline
 ArrayView<const T>::ArrayView(const T* p, size_type size_in, const ERCPNodeLookup rcpNodeLookup )
-  : ptr_(p), size_(size_in)
+  : ptr_(size_in == 0 ? nullptr : p), size_(size_in)
 {
 #ifdef HAVE_TEUCHOS_ARRAY_BOUNDSCHECK
-  TEUCHOS_TEST_FOR_EXCEPT( p != 0 && size_in <= 0 );
+  // We comment out the one check below, as part of the fix for #4234:
+  //
+  // https://github.com/trilinos/Trilinos/issues/4234
+  //
+  // This permits conversion from std::vector or Kokkos::View, using
+  // ArrayView(x.data(), x.size(), RCP_DISABLE_NODE_LOOKUP).  The
+  // other part of the fix is that we make sure ptr_ is null if
+  // size_in is zero.
+  //
+  //TEUCHOS_TEST_FOR_EXCEPT( p != 0 && size_in <= 0 );
+
   TEUCHOS_TEST_FOR_EXCEPT( p == 0 && size_in != 0 );
   // This only does something if HAVE_TEUCHOS_ARRAY_BOUNDSCHECK is defined.
   setUpIterators(rcpNodeLookup);

--- a/packages/teuchos/kokkoscompat/src/KokkosCompat_View.hpp
+++ b/packages/teuchos/kokkoscompat/src/KokkosCompat_View.hpp
@@ -185,8 +185,9 @@ namespace Kokkos {
       // allocated (e.g.,) using new or malloc, but it's not a problem
       // here, because the custom deallocator does not free anything.
       // Nevertheless, it's better not to trouble the tracking system.
-      return Teuchos::arcp(view.data(), 0, view.span(),
-                           deallocator(view), false);
+      return Teuchos::arcp (view.span () == 0 ? nullptr : view.data (),
+                            0, view.span (),
+                            deallocator(view), false);
     }
 
     // Create a "persisting view" from a Kokkos::View
@@ -195,9 +196,12 @@ namespace Kokkos {
     persistingView(
       const ViewType& view,
       typename Teuchos::ArrayRCP<typename ViewType::value_type>::size_type offset,
-      typename Teuchos::ArrayRCP<typename ViewType::value_type>::size_type size) {
-      return Teuchos::arcp(view.data()+offset, 0, size,
-                           deallocator(view), false);
+      typename Teuchos::ArrayRCP<typename ViewType::value_type>::size_type size)
+    {
+      // This check is related to #4234.
+      typename ViewType::value_type* const ptr =
+        (size == 0 && offset == 0) ? nullptr : (view.data () + offset);
+      return Teuchos::arcp (ptr, 0, size, deallocator (view), false);
     }
 
     template <class ViewType, typename Ordinal>

--- a/packages/teuchos/kokkoscompat/test/CMakeLists.txt
+++ b/packages/teuchos/kokkoscompat/test/CMakeLists.txt
@@ -13,7 +13,6 @@ TRIBITS_ADD_EXECUTABLE_AND_TEST(
   linkTest
   SOURCES
     linkTest.cpp 
-    ${TEUCHOS_STD_UNIT_TEST_MAIN}
   ARGS ""
   COMM serial mpi
   NUM_MPI_PROCS 1

--- a/packages/teuchos/kokkoscompat/test/linkTest.cpp
+++ b/packages/teuchos/kokkoscompat/test/linkTest.cpp
@@ -47,15 +47,7 @@
 #include <KokkosCompat_TMM.hpp>
 #include <Teuchos_Array.hpp>
 
-// CMakeList.txt is not set up for compiling Cuda
-// so choose an appropriate non-cuda device:
-#if defined( KOKKOS_ENABLE_CUDA )
-typedef Kokkos::HostSpace::execution_space TestDevice ;
-#else
-typedef Kokkos::DefaultExecutionSpace TestDevice ;
-#endif
-
-namespace {
+namespace TestTeuchosKokkosCompat {
 
   // Test functor, that fills the given 1-D vector in parallel.  The
   // functor is like APL's iota, except that it starts with an initial
@@ -64,18 +56,10 @@ namespace {
   template<class Device>
   class FillFunctor {
   public:
-    // Functors need a execution_space typedef.
-    typedef Device execution_space;
-
-    // Constructor accepts a View of a 1-D array.
     FillFunctor (const Kokkos::View<double*, Device>& vec) : vec_ (vec) {}
-
-    // Initialize the array.
-    KOKKOS_INLINE_FUNCTION
-    void operator () (int i) const {
+    KOKKOS_INLINE_FUNCTION void operator () (const int i) const {
       vec_[i] = 42.0 + static_cast<double> (i);
     }
-
   private:
     Kokkos::View<double*, Device> vec_;
   };
@@ -99,14 +83,12 @@ namespace {
     ViewType view_; // View that owns the memory
   };
 
-} // namespace (anonymous)
-
+} // namespace TestTeuchosKokkosCompat
 
 // Just test whether Teuchos memory management objects and Kokkos
 // Array Views can coexist in the same program.  This test does not
 // have the Teuchos and Kokkos objects interact with each other.
 TEUCHOS_UNIT_TEST( LinkTeuchosAndKokkos, NoInteraction ) {
-  TestDevice::initialize();
   typedef Teuchos::Array<double>::size_type size_type;
 
   const size_type numElts = 10;
@@ -120,38 +102,55 @@ TEUCHOS_UNIT_TEST( LinkTeuchosAndKokkos, NoInteraction ) {
     TEST_EQUALITY( xView[k], x[k+3] );
   }
 
-  typedef Kokkos::View<double*, TestDevice> ka_view_type;
+  typedef Kokkos::View<double*> ka_view_type;
   ka_view_type y ("y", numElts);
-  Kokkos::parallel_for (y.extent (0), FillFunctor<TestDevice> (y));
-  TestDevice::finalize();
+
+  using execution_space = ka_view_type::execution_space;
+  Kokkos::RangePolicy<execution_space, int> range (0, y.extent (0));
+  using functor_type = TestTeuchosKokkosCompat::FillFunctor<execution_space>;
+  Kokkos::parallel_for ("FillFunctor", range, functor_type (y));
 }
 
 
 // Get a Teuchos::ArrayView of a Kokkos::View, and make sure that
 // it points to the same data.
 TEUCHOS_UNIT_TEST( LinkTeuchosAndKokkos, ArrayViewOfView ) {
-  TestDevice::initialize();
   typedef Teuchos::Array<double>::size_type size_type;
-  typedef Kokkos::View<double*, Kokkos::LayoutLeft, TestDevice> ka_view_type;
+  typedef Kokkos::View<double*> ka_view_type;
 
   const size_type numElts = 10;
   ka_view_type y ("y", numElts);
-  Kokkos::parallel_for (y.extent (0), FillFunctor<TestDevice> (y));
 
-  double* const y_raw = y.data ();
-  const size_type y_size = static_cast<size_type> (y.extent (0));
+  using execution_space = ka_view_type::execution_space;
+  Kokkos::RangePolicy<execution_space, int> range (0, y.extent (0));
+  using functor_type = TestTeuchosKokkosCompat::FillFunctor<execution_space>;
+  Kokkos::parallel_for ("FillFunctor", range, functor_type (y));
 
-  Teuchos::ArrayView<double> y_view (y_raw, y_size);
-  TEST_EQUALITY_CONST( y_view.size(), y_size );
-  for (size_type k = 0; k < y_size; ++k) {
-    TEST_EQUALITY( y_view[k], y[k] );
+  auto y_host = Kokkos::create_mirror_view (y);
+  Kokkos::deep_copy (y_host, y);
+
+  Teuchos::ArrayView<double> y_view (y_host.data (), y_host.extent (0));
+  TEST_EQUALITY_CONST( size_t (y_view.size ()), size_t (y.extent (0)) );
+  if (success) {
+    for (size_type k = 0; k < size_type (y.extent (0)); ++k) {
+      TEST_EQUALITY( y_view[k], y[k] );
+    }
   }
-  TestDevice::finalize();
 }
 
+template<class DataType, class LayoutType = Kokkos::LayoutLeft>
+using ManagedHostView =
+  Kokkos::View<DataType, LayoutType,
+               Kokkos::Device<Kokkos::DefaultHostExecutionSpace, Kokkos::HostSpace> >;
 
-// Get a Kokkos::View of a Teuchos::ArrayView, and make sure that
-// it points to the same data.  Thanks to Christian Trott for
+template<class DataType, class LayoutType = Kokkos::LayoutLeft>
+using UnmanagedHostView =
+  Kokkos::View<DataType, LayoutType,
+               Kokkos::Device<Kokkos::DefaultHostExecutionSpace, Kokkos::HostSpace>,
+               Kokkos::MemoryTraits<Kokkos::Unmanaged> >;
+
+// Get a Kokkos::View of a Teuchos::ArrayView, and make sure that it
+// points to the same data.  Thanks to Christian Trott for
 // implementing the necessary functionality (View constructor for
 // certain View specializations, that takes a raw pointer and
 // dimensions) in Kokkos Array.
@@ -159,10 +158,7 @@ TEUCHOS_UNIT_TEST( LinkTeuchosAndKokkos, ArrayViewOfView ) {
 // This example will be useful for implementing the
 // Tpetra::MultiVector methods get1dCopy and get2dCopy.
 TEUCHOS_UNIT_TEST( LinkTeuchosAndKokkos, ViewOfArrayView ) {
-  TestDevice::initialize();
-  typedef Teuchos::Array<double>::size_type size_type;
-  typedef Kokkos::View<double*, Kokkos::LayoutLeft, TestDevice, Kokkos::MemoryUnmanaged> ka_view_type;
-  typedef Kokkos::View<const double*, Kokkos::LayoutLeft, TestDevice, Kokkos::MemoryUnmanaged> ka_const_view_type;
+  using size_type = Teuchos::Array<double>::size_type;
 
   const size_type numElts = 10;
   Teuchos::Array<double> x (numElts);
@@ -170,38 +166,31 @@ TEUCHOS_UNIT_TEST( LinkTeuchosAndKokkos, ViewOfArrayView ) {
     x[k] = 42.0 + static_cast<double> (k);
   }
   // You can make an (unmanaged) View of a raw array with left or
-  // right (Fortran or C) layout on the HostSpace device, just by passing
-  // the array and stride(s) into the constructor.  If you want the
-  // dimensions to differ from the strides, you'll have to create a
-  // subview with the desired dimensions.
-  ka_view_type x_view (x.getRawPtr (), x.size ());
+  // right (Fortran or C) layout on the HostSpace device, just by
+  // passing the array and stride(s) into the constructor.  If you
+  // want the dimensions to differ from the strides, you'll have to
+  // create a subview with the desired dimensions.
+  UnmanagedHostView<double*> x_view (x.getRawPtr (), x.size ());
 
-  TEST_EQUALITY( x.size(), static_cast<size_type>(x_view.extent(0)) );
+  TEST_EQUALITY( x.size(), static_cast<size_type> (x_view.extent (0)) );
   for (size_type k = 0; k < x.size (); ++k) {
-    TEST_EQUALITY( x_view[k], x[k] );
+    TEST_EQUALITY( x_view(k), x[k] );
   }
 
-  // This ensures that conversions from double* to const double* work correctly.
-  // x.getRawPtr() returns double*.
-  ka_const_view_type x_view_const ( (const double*) x.getRawPtr (), x.size ());
+  UnmanagedHostView<const double*> x_view_const (x.getRawPtr (), x.size ());
 
-  TEST_EQUALITY( x.size(), static_cast<size_type>(x_view_const.extent(0)) );
+  TEST_EQUALITY( x.size (), static_cast<size_type> (x_view_const.extent (0)) );
   for (size_type k = 0; k < x.size (); ++k) {
-    TEST_EQUALITY( x_view_const[k], x[k] );
+    TEST_EQUALITY( x_view_const(k), x[k] );
   }
-
-  TestDevice::finalize();
 }
-
 
 // Create a 2-D View (LayoutLeft, in HostSpace memory), and then create an
 // ArrayRCP which owns that View (using a custom destructor).  This
 // will be useful for implementing Tpetra::MultiVector's getData,
 // getDataNonConst, get1dView, and get1dViewNonConst methods.
 TEUCHOS_UNIT_TEST( LinkTeuchosAndKokkos, ArrayRCP1D_of_2DView ) {
-  TestDevice::initialize();
-
-  typedef Kokkos::View<double**, Kokkos::LayoutLeft, TestDevice> ka_view_type;
+  using ka_view_type = ManagedHostView<double**>;
 
   const size_t numRows = 75;
   const size_t numCols = 5;
@@ -209,7 +198,8 @@ TEUCHOS_UNIT_TEST( LinkTeuchosAndKokkos, ArrayRCP1D_of_2DView ) {
   const size_t ZERO = static_cast<size_t> (0);
 
   ka_view_type X ("X", stride, numCols);
-  ka_view_type X_view = Kokkos::subview (X, std::make_pair (ZERO, numRows), std::make_pair (ZERO, numCols));
+  ka_view_type X_view = Kokkos::subview (X, std::make_pair (ZERO, numRows),
+                                         std::make_pair (ZERO, numCols));
   TEST_EQUALITY(X_view.extent(0), numRows);
   TEST_EQUALITY(X_view.extent(1), numCols);
 
@@ -240,12 +230,12 @@ TEUCHOS_UNIT_TEST( LinkTeuchosAndKokkos, ArrayRCP1D_of_2DView ) {
   // View's raw pointer, but still defers to the View for memory
   // management.
 
-  Teuchos::ArrayRCP<double> Y_values (X.data (), 0, stride*numCols, Deallocator<double, ka_view_type> (X), true);
+  using dealloc_type = TestTeuchosKokkosCompat::Deallocator<double, ka_view_type>;
+  Teuchos::ArrayRCP<double> Y_values (X.data (), 0, stride*numCols,
+                                      dealloc_type (X), true);
   TEST_EQUALITY(Y_values.getRawPtr(), X.data());
   TEST_EQUALITY(Y_values.getRawPtr(), X_view.data());
   TEST_EQUALITY(Y_values.size(), stride*numCols);
-
-  TestDevice::finalize();
 }
 
 
@@ -262,8 +252,7 @@ TEUCHOS_UNIT_TEST( LinkTeuchosAndKokkos, ArrayRCP2D_of_2DView ) {
   // multivector still needs to be a View<double**, ...>, and we have
   // to use the subview() overload that takes two ranges of row and
   // column indices.
-  typedef Kokkos::View<double**, Kokkos::LayoutLeft, TestDevice> ka_view_type;
-  TestDevice::initialize();
+  using ka_view_type = ManagedHostView<double**>;
 
   const size_t numRows = 75;
   const size_t numCols = 5;
@@ -271,7 +260,8 @@ TEUCHOS_UNIT_TEST( LinkTeuchosAndKokkos, ArrayRCP2D_of_2DView ) {
   const size_t ZERO = static_cast<size_t> (0);
 
   ka_view_type X ("X", stride, numCols);
-  ka_view_type X_view = Kokkos::subview (X, std::make_pair (ZERO, numRows), std::make_pair (ZERO, numCols));
+  ka_view_type X_view = Kokkos::subview (X, std::make_pair (ZERO, numRows),
+                                         std::make_pair (ZERO, numCols));
   TEST_EQUALITY( & X(0,0) , & X_view(0,0) );
   TEST_EQUALITY(X_view.extent(0), numRows);
   TEST_EQUALITY(X_view.extent(1), numCols);
@@ -292,7 +282,8 @@ TEUCHOS_UNIT_TEST( LinkTeuchosAndKokkos, ArrayRCP2D_of_2DView ) {
   // will implement Tpetra::MultiVector methods like get2dView.
   Teuchos::ArrayRCP<Teuchos::ArrayRCP<double> > Y_2D (X_view.extent (1));
   for (size_t j = 0; j < static_cast<size_t> (X_view.extent (1)); ++j) {
-    ka_view_type X_j = Kokkos::subview (X_view, std::make_pair (ZERO, numRows), std::make_pair (j, j+1));
+    ka_view_type X_j = Kokkos::subview (X_view, std::make_pair (ZERO, numRows),
+                                        std::make_pair (j, j+1));
     TEST_EQUALITY( & X_view(0,j) , & X_j(0,0) );
     TEST_EQUALITY(static_cast<size_t>(X_j.extent(0)), numRows);
     TEST_EQUALITY_CONST(X_j.extent(1), 1);
@@ -316,10 +307,22 @@ TEUCHOS_UNIT_TEST( LinkTeuchosAndKokkos, ArrayRCP2D_of_2DView ) {
     // X_j, and doesn't actually deallocate memory.  That way, the
     // ArrayRCP can use the View's raw pointer, but still defers to
     // the View for memory management.
-    Teuchos::ArrayRCP<double> Y_j (X_j.data (), 0, numRows, Deallocator<double, ka_view_type> (X_j), true);
+    using dealloc_type = TestTeuchosKokkosCompat::Deallocator<double, ka_view_type>;
+    Teuchos::ArrayRCP<double> Y_j (X_j.data (), 0, numRows,
+                                   dealloc_type (X_j), true);
     Y_2D[j] = Y_j;
   }
-  TestDevice::finalize();
 }
 
 
+int
+main (int argc, char* argv[])
+{
+  // Initialize MPI (if enabled) before initializing Kokkos.  This
+  // lets MPI control things like pinning processes to sockets.
+  Teuchos::GlobalMPISession mpiSession (&argc, &argv);
+  Kokkos::ScopeGuard kokkosScope (argc, argv);
+  const int errCode =
+    Teuchos::UnitTestRepository::runUnitTestsFromMain (argc, argv);
+  return errCode;
+}

--- a/packages/teuchos/numerics/test/LAPACK/cxx_main.cpp
+++ b/packages/teuchos/numerics/test/LAPACK/cxx_main.cpp
@@ -62,7 +62,7 @@ int main(int argc, char* argv[])
 
   // Create some common typedefs
   typedef Teuchos::ScalarTraits<double> STS;
-  typedef Teuchos::ScalarTraits<double>::magnitudeType MagnitudeType;
+  typedef STS::magnitudeType MagnitudeType;
   typedef Teuchos::ScalarTraits<MagnitudeType> STM;
 
   Teuchos::LAPACK<int,double> L;
@@ -89,8 +89,8 @@ int main(int argc, char* argv[])
         if (verbose && i==3) std::cout << "passed!" << std::endl;
       } else {
         if (verbose) std::cout << "FAILED" << std::endl;
-        numberFailedTests++;	
-	break;
+        numberFailedTests++;
+        break;
       }
     }
 
@@ -128,13 +128,13 @@ int main(int argc, char* argv[])
     numberFailedTests++;
   }
   else
-  { 
+  {
     for (int i=0; i<n_potrf; i++)
     {
-      if ( diag_a[i*n_potrf + i] == (i+1) ) 
+      if ( diag_a[i*n_potrf + i] == (i+1) )
       {
         if (verbose && i==(n_potrf-1)) std::cout << "passed!" << std::endl;
-      } 
+      }
       else
       {
         if (verbose) std::cout << "FAILED" << std::endl;
@@ -145,11 +145,11 @@ int main(int argc, char* argv[])
   }
 
   if (verbose) std::cout << "POCON test ... ";
-  
+
   double anorm = (n_potrf*n_potrf), rcond;
   std::vector<double> work(3*n_potrf);
   std::vector<int> iwork(n_potrf);
-  
+
   L.POCON(char_U, n_potrf, &diag_a[0], n_potrf, anorm, &rcond, &work[0], &iwork[0], &info);
   if (info != 0 || (rcond != 1.0/anorm))
   {
@@ -157,13 +157,13 @@ int main(int argc, char* argv[])
     if (verbose) std::cout << "FAILED" << std::endl;
   }
   else
-  { 
+  {
     if (verbose) std::cout << "passed!" << std::endl;
-  } 
+  }
 
   if (verbose) std::cout << "POTRI test ... ";
-  std::vector<double> diag_a_trtri(diag_a); // Save a copy for TRTRI test 
-   
+  std::vector<double> diag_a_trtri(diag_a); // Save a copy for TRTRI test
+
   L.POTRI(char_U, n_potrf, &diag_a[0], n_potrf, &info);
 
   if (info != 0 || (diag_a[n_potrf+1] != 1.0/4.0))
@@ -172,20 +172,20 @@ int main(int argc, char* argv[])
     if (verbose) std::cout << "FAILED" << std::endl;
   }
   else
-  { 
+  {
     if (verbose) std::cout << "passed!" << std::endl;
-  } 
- 
+  }
+
   if (verbose) std::cout << "TRTRI test ... ";
-  
+
   int n_trtri = n_potrf;
   L.TRTRI( char_U, char_N, n_trtri, &diag_a_trtri[0], n_trtri, &info );
   for (int i=0; i<n_trtri; i++)
   {
-    if ( diag_a_trtri[i*n_trtri + i] == 1.0/(i+1) ) 
+    if ( diag_a_trtri[i*n_trtri + i] == 1.0/(i+1) )
     {
       if (verbose && i==(n_trtri-1)) std::cout << "passed!" << std::endl;
-    } 
+    }
     else
     {
       if (verbose) std::cout << "FAILED" << std::endl;
@@ -211,7 +211,7 @@ int main(int argc, char* argv[])
   }
 
 #endif
- 
+
   if (verbose) std::cout << "STEQR test ... ";
 
 #ifndef TEUCHOSNUMERICS_DISABLE_STEQR_TEST
@@ -235,8 +235,8 @@ int main(int argc, char* argv[])
   if (info != 0)
   {
     if (verbose)  std::cout << "STEQR: compute symmetric tridiagonal eigenvalues: "
-		  << "LAPACK's _STEQR failed with info = "
-		  << info;
+                  << "LAPACK's _STEQR failed with info = "
+                  << info;
 
       numberFailedTests++;
   }
@@ -249,8 +249,8 @@ int main(int argc, char* argv[])
   if ((fabs(lambda_min-exp_lambda_min)<1e-12) && (fabs(lambda_max-exp_lambda_max)<1e-12))
   {
     if (verbose) std::cout << "passed!" << std::endl;
-  } 
-  else 
+  }
+  else
   {
     if (verbose) std::cout << "FAILED" << std::endl;
     numberFailedTests++;

--- a/packages/tpetra/core/src/Tpetra_MultiVector_def.hpp
+++ b/packages/tpetra/core/src/Tpetra_MultiVector_def.hpp
@@ -62,13 +62,11 @@
 #include "Tpetra_Details_reallocDualViewIfNeeded.hpp"
 #include "Tpetra_Details_PackTraits.hpp"
 #include "Tpetra_KokkosRefactor_Details_MultiVectorDistObjectKernels.hpp"
-
-
-
 #include "KokkosCompat_View.hpp"
 #include "KokkosBlas.hpp"
 #include "KokkosKernels_Utils.hpp"
 #include "Kokkos_Random.hpp"
+#include "Kokkos_ArithTraits.hpp"
 
 #ifdef HAVE_TPETRA_INST_FLOAT128
 namespace Kokkos {
@@ -299,6 +297,43 @@ namespace { // (anonymous)
 
 
 namespace Tpetra {
+
+  namespace Details {
+    // Work-around for #3823.  The right way to fix this is to fix
+    // KokkosBlas::dot for complex, but this at least makes Tpetra's
+    // tests pass for complex.
+    template<class XVector,class YVector>
+    typename ::Kokkos::Details::InnerProductSpaceTraits<typename XVector::non_const_value_type>::dot_type
+    localDotWorkAround (const XVector& x, const YVector& y)
+    {
+      using x_value_type = typename XVector::non_const_value_type;
+      using y_value_type = typename YVector::non_const_value_type;
+
+      if (Kokkos::ArithTraits<x_value_type>::is_complex ||
+          Kokkos::ArithTraits<y_value_type>::is_complex) {
+        using IPT = ::Kokkos::Details::InnerProductSpaceTraits<x_value_type>;
+        using dot_type = typename IPT::dot_type;
+        using execution_space = typename XVector::execution_space;
+        using range_type = Kokkos::RangePolicy<execution_space, int>;
+        // Use double precision internally; this should improve
+        // accuracy for complex<float> and thus help more tests pass.
+        using impl_dot_type = typename Teuchos::ScalarTraits<dot_type>::doublePrecision;
+
+        impl_dot_type result;
+        Kokkos::parallel_reduce
+          ("Tpetra::MultiVector oneColDotWorkAround",
+           range_type (0, x.extent (0)),
+           KOKKOS_LAMBDA (const int lclRow, impl_dot_type& dst) {
+            dst += IPT::dot (x(lclRow), y(lclRow));
+          }, result);
+        return static_cast<dot_type> (result);
+      }
+      else {
+        return KokkosBlas::dot (x, y);
+      }
+    }
+  } // namespace Details
+
 
   template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
   bool
@@ -1904,7 +1939,8 @@ namespace Tpetra {
         auto x_1d = Kokkos::subview (x_2d, rowRng, 0);
         auto y_2d = y.template getLocalView<dev_memory_space> ();
         auto y_1d = Kokkos::subview (y_2d, rowRng, 0);
-        lclDot = KokkosBlas::dot (x_1d, y_1d);
+        // Work-around for #3823; see notes above.
+        lclDot = ::Tpetra::Details::localDotWorkAround (x_1d, y_1d);
 
         if (x.isDistributed ()) {
           using Teuchos::outArg;

